egs/librispeech/asr/nnlm/main.py

```diff
@@ -0,0 +1,331 @@
+    if args.model != 'Transformer':
+        hidden = model.init_hidden(batch_size)
+    for batch_idx in range(0, len(train_data) // batch_size - 1):
+        data, targets = get_batch(train_data, batch_idx)
+        # Starting each batch, we detach the hidden state from how it was previously produced.
+        # If we didn't, the model would try backpropagating all the way to start of the dataset.
+        model.zero_grad()
+        if args.model == 'Transformer':
+            output = model(data)
+            output = output.view(-1, ntokens)
+        else:
+            hidden = repackage_hidden(hidden)
+            output, hidden = model(data, hidden)
+
+        # import pdb; pdb.set_trace()
+        loss = criterion(output, targets)
+        loss.backward()
+
+        # `clip_grad_norm` helps prevent the exploding gradient problem in RNNs / LSTMs.
+        torch.nn.utils.clip_grad_norm_(model.parameters(), args.clip)
+        for p in model.parameters():
+            p.data.add_(p.grad, alpha=-lr)
+
+        #import pdb; pdb.set_trace()
+        total_loss += loss.item()
+
+        if batch_idx % args.log_interval == 0 and batch_idx > 0:
+            cur_loss = total_loss / args.log_interval
+            elapsed = time.time() - start_time
+            print(
+                '| epoch {:3d} | {:5d}/{:5d} batches | lr {:02.2f} | ms/batch {:5.2f} | '
+                'loss {:5.2f} | ppl {:8.2f}'.format(
+                    epoch, batch_idx,
+                    len(train_data) // batch_size, lr,
+                    elapsed * 1000 / args.log_interval, cur_loss,
+                    math.exp(cur_loss)))
+            total_loss = 0
+            start_time = time.time()
+        if args.dry_run:
+            break
+
+
+def export_onnx(path, batch_size, seq_len):
+    print('The model is also exported in ONNX format at {}'.format(
+        os.path.realpath(args.onnx_export)))
+    model.eval()
+    dummy_input = torch.LongTensor(seq_len * batch_size).zero_().view(
+        -1, batch_size).to(device)
+    hidden = model.init_hidden(batch_size)
+    torch.onnx.export(model, (dummy_input, hidden), path)
+
+
+# Loop over epochs.
+lr = args.lr
+best_val_loss = None
+
+# At any point you can hit Ctrl + C to break out of training early.
+try:
+    for epoch in range(1, args.epochs + 1):
+        epoch_start_time = time.time()
+        train()
+        val_loss = evaluate(val_data)
+        print('-' * 89)
+        print('| end of epoch {:3d} | time: {:5.2f}s | valid loss {:5.2f} | '
+              'valid ppl {:8.2f}'.format(epoch,
+                                         (time.time() - epoch_start_time),
+                                         val_loss, math.exp(val_loss)))
+        print('-' * 89)
+        # Save the model if the validation loss is the best we've seen so far.
+        if not best_val_loss or val_loss < best_val_loss:
+            with open(args.save, 'wb') as f:
+                torch.save(model, f)
```

Comment:
From https://pytorch.org/tutorials/beginner/saving_loading_models.html

> The disadvantage of this approach is that the serialized data is bound to the specific classes and the exact directory structure used when the model is saved.

Could you save only the state dict of the model?

Reply:
[solved](https://github.com/glynpu/snowfall/blob/nnlm/egs/librispeech/asr/nnlm/local/common.py#L30) as following
```
def save_checkpoint(filename: Pathlike,
                    model: torch.nn.Module,
                    info: Info = None) -> None:
    if not os.path.exists(os.path.dirname(filename)):
        Path(os.path.dirname(filename)).mkdir(parents=True, exist_ok=True)
    logging.info(f'Save checkpoint to {filename}')
    checkpoint = {
        'state_dict': model.state_dict(),
    }
    if info is not None:
        checkpoint.update(info)

    torch.save(checkpoint, filename)

```